### PR TITLE
Add generic `NonZero` type.

### PR DESCRIPTION
**Pull request by reitermarkus**
Add generic `NonZero` type as suggested in https://github.com/rust-lang/rust/issues/82363.

Currently, only the type itself is generic, the implementation is still done individually for each type with the existing macros.

It should be possible to implement most methods generically, but this would be incompatible with the stability attributes since unsigned and signed non-zero types were stabilized in different versions. Not sure what the best way to deal with this is.

r? @joshtriplett

### CODE REVIEW

**Comment by CAD97** [comment]
> incompatible with the stability attributes since unsigned and signed non-zero types were stabilized in different versions

Version numbers on `#[stable]` attributes are informational, so it'd be fine to adjust the method stability attributes to the earlier version, so long as the same set of methods are stable on all of the types. Providing a later `#[stable]` attribute on the generic type and accurate `#[stable]` on the `use`/`type` aliases should be sufficient.

Of note, though, is that changing `pub struct [<NonZero$int:upper>]($int)` to `pub type <NonZero$int:upper>] = NonZero<$int>` *is* technically a breaking change, as I can write `NonZeroUsize::<>` before but not after the change.

**Comment by leonardo-m** [comment]
This simplifies my code.

**Comment by leonardo-m** [comment]
This becomes useless then?
https://github.com/rust-lang/rust/issues/84352

**Comment by MomoLangenstein** [comment]
I am really excited about this! One quick question - why is `ZeroablePrimitive` implemented for `*mut T` but not for `*const T`?

**Comment by reitermarkus** [comment]
@MomoLangenstein, `NonNull<T>` is constructed with a `*mut T` but stores a `*const T`. I'm not yet sure though if `NonNull<T>` should be implemented via `NonZero<*const T>`.

**Comment by MomoLangenstein** [comment]
@reitermarkus Thanks for the explanation! Would it be useful to also allow wrapping `*const T` inside `NonZero`?

**Comment by reitermarkus** [comment]
> Would it be useful to also allow wrapping `*const T` inside `NonZero`?

I don't think so, a non-zero `*const T` is usually just a `&T` and if it's pointing to uninitialized data, it has to go through `*mut T` to initialize it. There may of course be a use-case I am missing.

File: src/test/ui/binop/issue-28837.rs [hunk 1]
```diff
@@ -15,7 +15,9 @@ fn main() {
 
     a & a; //~ ERROR no implementation for `A & A`
 
-    a | a; //~ ERROR no implementation for `A | A`
+    a | a;
+    //~^ ERROR no implementation for `A | NonZero<A>`
```
Comment (reitermarkus) [inline src/test/ui/binop/issue-28837.rs:19]: I adjusted this test for now so is passes, but this seems unrelated. Especially 

```
must implement `core::num::nonzero::ZeroablePrimitive`
```

in the output should not be suggested.

File: src/test/ui/or-patterns/or-patterns-syntactic-fail.rs [hunk 2]
```diff
@@ -8,7 +8,9 @@ use E::*;
 
 fn no_top_level_or_patterns() {
     // We do *not* allow or-patterns at the top level of lambdas...
-    let _ = |A | B: E| (); //~ ERROR no implementation for `E | ()`
+    let _ = |A | B: E| ();
+    //~^ ERROR mismatched types
+    //~^^ ERROR no implementation for `E | NonZero<E>`
```
Comment (reitermarkus) [inline src/test/ui/or-patterns/or-patterns-syntactic-fail.rs:13]: Same goes for this, but this seems to be cause by the same problem related to `BitOr`.

**Comment by scottmcm** [comment]
Note that a generic type was how these were implemented back in 1.26: <https://doc.rust-lang.org/1.26.0/src/core/num/mod.rs.html#53>

It might be good to look at why that changed to the current form before changing back again.

**Comment by reitermarkus** [comment]
According to https://github.com/rust-lang/rfcs/pull/2307, the concrete types were introduced to avoid the problem of what should happen if `Zeroable` is implemented for arbitrary types which are not pointers or integers. This problem does not exist here since the trait in this PR is sealed.

This is also explained by @kennytm here: https://github.com/rust-lang/rust/issues/82363#issuecomment-921513910

File: library/core/src/num/nonzero.rs [hunk 3]
```diff
@@ -1,200 +1,472 @@
...
-            }
+#[unstable(feature = "generic_nonzero", issue = "82363")]
+impl<T: ZeroablePrimitive> const PartialEq for NonZero<T>
+where
+    T: ~const PartialEq,
+{
+    #[inline]
+    fn eq(&self, other: &Self) -> bool {
+        // SAFETY: A `NonZero` is guaranteed to only contain non-zero values.
+        unsafe { self.0 == other.0 }
```
Comment (joshtriplett) [inline library/core/src/num/nonzero.rs:138]: This doesn't seem like it needs an unsafe block.

File: library/core/src/num/nonzero.rs [hunk 4]
```diff
@@ -1,200 +1,472 @@
...
-                    // SAFETY: since `rhs` is nonzero, the result of the
-                    // bitwise-or will be nonzero regardless of the value of
-                    // `self`.
-                    unsafe { $Ty::new_unchecked(self | rhs.get()) }
-                }
-            }
+    #[inline]
+    fn ne(&self, other: &Self) -> bool {
+        // SAFETY: A `NonZero` is guaranteed to only contain non-zero values.
+        unsafe { self.0 != other.0 }
```
Comment (joshtriplett) [inline library/core/src/num/nonzero.rs:144]: This doesn't seem like it needs an unsafe block.

File: library/core/src/num/nonzero.rs [hunk 5]
```diff
@@ -1,200 +1,472 @@
...
-            impl const BitOrAssign<$Int> for $Ty {
+#[unstable(feature = "generic_nonzero", issue = "82363")]
+impl<T: ZeroablePrimitive> const PartialOrd for NonZero<T>
+where
+    T: ~const PartialOrd,
+{
+    #[inline]
+    fn partial_cmp(&self, other: &Self) -> Option<Ordering> {
+        // SAFETY: A `NonZero` is guaranteed to only contain non-zero values.
+        unsafe { self.0.partial_cmp(&other.0) }
```
Comment (joshtriplett) [inline library/core/src/num/nonzero.rs:160]: This and the rest of this impl doesn't seem like it needs any unsafe blocks.

**Comment by joshtriplett** [comment]
Posted some feedback. This seems generally reasonable to me, though.

File: library/core/src/num/nonzero.rs (hunk 5 above)
Comment (reitermarkus) [reply library/core/src/num/nonzero.rs:160]: These do need `unsafe`, otherwise I get this error:

```
borrow of layout constrained field with interior mutability is unsafe and requires unsafe block
```

Comment (reitermarkus) [reply library/core/src/num/nonzero.rs:160]: Which is also why these traits cannot be derived and has to be implemented manually.

**Comment by anden3** [comment]
Hello @reitermarkus! I saw that this PR has some merge conflicts, what's the status of it? :)

**Comment by reitermarkus** [comment]
Still waiting for someone to review it.

**Comment by anden3** [comment]
> Still waiting for someone to review it.

Sure, that's fine! It's just much easier for someone to review it if it doesn't have merge conflicts :)

File: library/core/src/mem/mod.rs [hunk 6]
```diff
@@ -1168,7 +1168,7 @@ impl<T> fmt::Debug for Discriminant<T> {
 /// [`as`]: ../../std/keyword.as.html
 /// [primitive representation]: ../../reference/type-layout.html#primitive-representations
 /// [default representation]: ../../reference/type-layout.html#the-default-representation
-/// ```
+/// ```ignore
```
Comment (reitermarkus) [inline library/core/src/mem/mod.rs:1171]: `*const _` below cannot be resolved for some reason, no idea what causes this.

File: src/tools/clippy/tests/ui/transmute_int_to_non_zero.stderr [hunk 7]
```diff
@@ -2,63 +2,63 @@ error: transmute from a `u8` to a `NonZeroU8`
   --> $DIR/transmute_int_to_non_zero.rs:18:33
    |
 LL |     let _: NonZeroU8 = unsafe { std::mem::transmute(int_u8) };
-   |                                 ^^^^^^^^^^^^^^^^^^^^^^^^^^^ help: consider using: `NonZeroU8::new_unchecked(int_u8)`
+   |                                 ^^^^^^^^^^^^^^^^^^^^^^^^^^^ help: consider using: `NonZero::new_unchecked(int_u8)`
```
Comment (flip1995) [inline src/tools/clippy/tests/ui/transmute_int_to_non_zero.stderr:5]: Since this feature will be unstable, we should not suggest `NonZero` without gating on the feature in Clippy. This test file should stay unchanged, until `NonZero` is stabilized. You might want to add another test file with the feature enabled to test, if the generic `NonZero::...` suggestion is produced, but that's optional.

**Review by flip1995 — CHANGES_REQUESTED** [review]
Clippy comments

File: src/tools/clippy/tests/ui/transmute_int_to_non_zero.stderr (hunk 7 above)
Comment (reitermarkus) [reply src/tools/clippy/tests/ui/transmute_int_to_non_zero.stderr:5]: 1. How to gate on the feature?
2. Alternatively, is there any way of detecting whether a `rust_middle::ty::Ty` is a type alias?

**Comment by reitermarkus** [comment]
@joshtriplett, could you review this again? Thanks.

File: src/tools/clippy/tests/ui/transmute_int_to_non_zero.stderr (hunk 7 above)
Comment (flip1995) [reply src/tools/clippy/tests/ui/transmute_int_to_non_zero.stderr]: You can check over the `cx.sess().features.<feature_name>` whether the feature is enabled or not.

You can probably check for the `ty.kind() == ty::Alias(..)`: https://doc.rust-lang.org/nightly/nightly-rustc/rustc_type_ir/sty/enum.TyKind.html#variant.Alias

**Review by flip1995 — APPROVED** [review]
Clippy changes LGTM. Consider adding a `FIXME` comment, that this code can be potentially simplified once the feature is stable.

**Comment by jendrikw** [comment]
> I don't think so, a non-zero `*const T` is usually just a `&T` and if it's pointing to uninitialized data, it has to go through `*mut T` to initialize it. There may of course be a use-case I am missing.

A possible usecase might be unaligned pointers.

**Comment by tgross35** [comment]
@dtolnay are you waiting on @reitermarkus to rebase this for reviewing? Seems like the bits keep rotting here :(

**Review by dtolnay — CHANGES_REQUESTED** [review]
Thanks for the PR!

All the parts outside of nonzero.rs look great to me.

Nonzero.rs is challenging to review in this PR because all of the following are being done simultaneously in that single commit:

- expanding derived impls into handwritten impls
- modifying non-generic associated functions into generic ones
- refactoring macro-generated trait impls
- introducing brand new functionality that wasn't available in the standard library before, namely something to do with nonzero raw pointers, along with new From impls
- changing code inside function bodies, such as the `cfg(debug_assertions)` which appeared inside `new_unchecked`
- adding the new struct definition and aliases
- reordering some big blocks of code
- adapting documentation of stable methods to use unstable `NonZero<T>` instead of the original stable types

Overall I feel that I couldn't review this as a diff with sufficient confidence, only by reviewing the end result from scratch. And reviewing the end result, it's unlikely that I would be able to pick up on anything that has been inadvertently added or gone missing, especially as this is the result of some gnarly rebases, it sounds like.

Regarding the raw pointer stuff, I am not interested in merging that in this PR. FWIW that is what is causing that doctest on core::mem::discriminant to fail. There has been a From impl added which didn't exist before this PR, and is causing an inference breakage.

The doc changes on methods like `checked_add` also shouldn't merge in this PR, I think. `NonZero<T>` is going to be nightly-only for a period of at least a few months. During that time, it's extra important that the documentation shows how to continue using the existing stable functionality using only the stable aliases.

Separately, I noticed from building the documentation with `x.py doc library/core`, this ends up being an unexpectedly significant regression in usability of the docs for NonZero, since there end up being 67 separate impl blocks on `NonZero<T>`. See the screenshots in #118665.

I think a sequence of reviewable changes to get `NonZero<T>` merged would be:

1. #118665 so that we don't get the 67 impl blocks in the docs

2. A self-contained change which introduces _private_ type aliases by doing `pub(crate) type NonZero<T> = <T as ZeroablePrimitive>::NonZero`. Note that the type alias here is in the opposite direction as the desired end state, however all the rest of the changes are in the correct direction (replacing code that mentions `NonZeroI8` with `NonZero<i8>`, such as in `core::convert`, and various macros that need to change `$:ident` to `$:ty`) so overall this ends up being good forward progress that can land independently (or even in multiple pieces).

3. Expand derived impls into handwritten impls in all the cases where they would otherwise hit _"borrow of layout constrained field with interior mutability"_ in step 4. This can land independently.

4. Add the new generic NonZero struct definition, delete the alias from step 2, and add non-generic aliases in the opposite direction. Because of steps 2 and 3, very little standard library code should need to change here. The clippy change would need to go here I think. Also this commit would insert some `unsafe` blocks into the handwritten impls from step 3, but no other changes. Maybe this would be about 1/10 the size of this PR.

5. Later during stabilization process: update the example code in docs to use `NonZero<T>`.

I have intentionally left out some of the changes I noted seeing in this PR. In particular:

- Converting non-generic associated functions into generic ones is not something that needs to happen in order to merge or stabilize NonZero. If a function is in `impl NonZeroI8` today, it should be in `impl NonZero<i8>` after this work, not in `impl<T> NonZero<T>`. That can happen afterward as an independent change, probably after stabilization of NonZero, or in the run-up to stabilization if there is a reason to think there would be a backward compatibility risk in doing it afterward.

- Same thing for trait impls. All the same trait impls should exist before and after the refactor. There shouldn't be new generic `impl<T> ... for NonZero<T>`. As with associated functions, the non-generic trait impls can be converted over to generic ones separately afterward.

**Comment by reitermarkus** [comment]
@dtolnay, I have added the private type alias here: https://github.com/rust-lang/rust/pull/119990

**Comment by reitermarkus** [comment]
@dtolnay, I have added the handwritten trait implementations here: https://github.com/rust-lang/rust/pull/120160
